iquip/apps/dataviewer.py

```diff
@@ -820,69 +810,127 @@ class _DatasetFetcherThread(QThread):
-            logger.exception("Failed to fetch the dataset %s.", self.name)
-            return
-        dataset = np.array(response.json())
-        try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": f"{self.name}.parameters"},
-                                    timeout=10)
-            response.raise_for_status()
-        except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset parameters.")
-            parameters = list(map(str, range(dataset.shape[1] - 1)))
-        else:
-            parameters = response.json()
+    def _get(
+        self,
+        path: str,
+        params: Dict[str, Any],
+        default: Any = None,
+        timeout: float = 10
+    ) -> Any:
+        """Returns the json()-ed response of a GET request.
+        
+        Args:
+            path: The API path, i.e., the request url is "http://{ip}:{port}/{path}".
+            params: Params argument for the GET request.
+            default: The return value of this method when an exception occurs
+              during the GET request.
+            timeout: Timeout argument for the GET request.
+        """
+        url = f"http://{self.ip}:{self.port}/{path}"
         try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": f"{self.name}.units"},
-                                    timeout=10)
+            response = requests.get(url, params=params, timeout=timeout)
             response.raise_for_status()
         except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset units.")
-            units = [None] * (dataset.shape[1] - 1)
+            logger.exception("Failed to GET %s", url)
+            return default
+        return response.json()
+
+    def _initialize(self) -> float:
+        """Fetches the target dataset to initialize the local dataset.
+        
+        Returns:
+            The received timestamp or -1 if it failed.
+        """
+        response = self._get("dataset/master/", {"key": self.name})
+        if response is None or response[0] < 0:
+            return -1
+        timestamp, rawDataset = response
+        dataset = np.array(rawDataset)
+        numberOfParameters = dataset.shape[1] if dataset.ndim > 1 else 0
+        _, parameters = self._get(
+            "dataset/master/",
+            {"key": f"{self.name}.parameters"},
+            (0, list(map(str, range(numberOfParameters)))),
+        )
+        rawUnits = self._get("dataset/master/", {"key": f"{self.name}.units"})
+        if rawUnits is None:
+            units = [None] * (numberOfParameters)
         else:
-            units = [unit if unit else None for unit in response.json()]
-        self.fetched.emit(dataset, parameters, units)
+            units = [unit if unit else None for unit in rawUnits[1]]
```

Comment:
Is it different from `units = rawUnits[1]`?

Reply:
Since artiq only supports single-type lists, the unit list only has strings.
So no-unit is `""` instead of `None`.

Reply:
Also, since the current version has a problem, this might be updated soon.
The problem is that a user should wait for up to 12 seconds after clicking the stop sync button.